File: src/Shared/Polyrific.Catapult.Shared.ApiClient/ExternalServiceService.cs
// Copyright (c) Polyrific, Inc 2018. All rights reserved.

using System.Collections.Generic;
using System.Threading.Tasks;
using Polyrific.Catapult.Shared.Dto.ExternalService;
using Polyrific.Catapult.Shared.Service;

namespace Polyrific.Catapult.Shared.ApiClient
{
    public class ExternalServiceService : BaseService, IExternalServiceService
    {
        public ExternalServiceService(IApiClient api) : base(api)
        {
        }

        public async Task<ExternalServiceDto> CreateExternalService(CreateExternalServiceDto dto)
        {
            var path = "service";

            return await Api.Post<CreateExternalServiceDto, ExternalServiceDto>(path, dto);
        }

        public async Task DeleteExternalService(int serviceId)
        {
            var path = $"service/{serviceId}";

            await Api.Delete(path);
        }

        public async Task<ExternalServiceDto> GetExternalService(int serviceId)
        {
            var path = $"service/{serviceId}";

            return await Api.Get<ExternalServiceDto>(path);
        }

        public async Task<ExternalServiceDto> GetExternalServiceByName(string name)
        {
            var path = $"service/name/{name}";

            return await Api.Get<ExternalServiceDto>(path);
        }

        public async Task<List<ExternalServiceDto>> GetExternalServices()
        {
            var path = $"service";

            return await Api.Get<List<ExternalServiceDto>>(path);
        }

        public async Task UpdateExternalService(int serviceId, UpdateExternalServiceDto dto)
        {
            var path = $"service/{serviceId}";

            await Api.Put<UpdateExternalServiceDto>(path, dto);
        }
    }
}

File: src/Web/opencatapultweb/src/app/core/services/external-service.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { Observable, Subject, BehaviorSubject, of } from 'rxjs';
import { ExternalServiceDto } from '../models/external-service/external-service-dto';
import { UpdateExternalServiceDto } from '../models/external-service/update-external-service-dto';
import { CreateExternalServiceDto } from '../models/external-service/create-external-service-dto';
import { concatMap } from 'rxjs/operators';

@Injectable()
export class ExternalServiceService {
  private currentExternalServices = new BehaviorSubject<ExternalServiceDto[]>([]);
  public externalServices: Observable<ExternalServiceDto[]> = this.currentExternalServices.asObservable();

  constructor(private api: ApiService) { }

  getExternalServiceByName(name: string): Observable<ExternalServiceDto> {
    return this.api.get<ExternalServiceDto>(`service/name/${name}`);
  }

  getExternalService(id: number) {
    return this.api.get<ExternalServiceDto>(`service/${id}`);
  }

  getExternalServices() {
    return this.api.get<ExternalServiceDto[]>(`service`)
      .pipe(concatMap((data) => {
        this.currentExternalServices.next(data);
        return of(data);
      }));
  }

  createExternalService(dto: CreateExternalServiceDto) {
    return this.api.post(`service`, dto);
  }

  updateExternalService(externalServiceId: number, dto: UpdateExternalServiceDto) {
    return this.api.put(`service/${externalServiceId}`, dto);
  }

  deleteExternalService(externalServiceId: number) {
    return this.api.delete(`service/${externalServiceId}`);
  }
}
